ENH: add 'origin' and 'offset' arguments to 'resample' and 'pd.Grouper'

hasB4K opened the pull request:
**EDIT:** this PR has changed, now instead of adding `adjust_timestamp` we are adding `origin` and `offset` arguments to `resample` and `pd.Grouper` (see https://github.com/pandas-dev/pandas/pull/31809#issuecomment-583884772)

----

Hello,

This enhancement is an alternative to the `base` argument present in `pd.Grouper` or in the method `resample`. It adds the `adjust_timestamp` argument to change the current behavior of: https://github.com/pandas-dev/pandas/blob/master/pandas/core/resample.py#L1728

-  `adjust_timestamp` is the timestamp on which to adjust the grouping. If None is passed, the first day of the time series at midnight is used.

Currently the bins of the grouping are adjusted based on the beginning of the day of the time series starting point. This works well with frequencies that are multiples of a day (like `30D`) or that divides a day (like `90s` or `1min`). But it can create inconsistencies with some frequencies that do not meet this criteria.

Here is a simple snippet from a test that I added that proves that the current behavior can lead to some inconsistencies. Inconsistencies that can be fixed if we use `adjust_timestamp`:

```python
import pandas as pd
import numpy as np
import pandas._testing as tm
import pytest


freq = "1399min"  # prime number that is smaller than 24h
start, end = "1/1/2000 00:00:00", "1/31/2000 00:00"
middle = "1/15/2000 00:00:00"

rng = pd.date_range(start, end, freq="1231min")  # prime number
ts = pd.Series(np.random.randn(len(rng)), index=rng)
ts2 = ts[middle:end]

# proves that grouper without a fixed adjust_timestamp does not work
# when dealing with unusual frequencies
simple_grouper = pd.Grouper(freq=freq)
count_ts = ts.groupby(simple_grouper).agg("count")
count_ts = count_ts[middle:end]
count_ts2 = ts2.groupby(simple_grouper).agg("count")
with pytest.raises(AssertionError):
    tm.assert_index_equal(count_ts.index, count_ts2.index)

# test adjusted_timestamp on 1970-01-01 00:00:00
adjust_timestamp = pd.Timestamp(0)
adjusted_grouper = pd.Grouper(freq=freq, adjust_timestamp=adjust_timestamp)
adjusted_count_ts = ts.groupby(adjusted_grouper).agg("count")
adjusted_count_ts = adjusted_count_ts[middle:end]
adjusted_count_ts2 = ts2.groupby(adjusted_grouper).agg("count")
tm.assert_series_equal(adjusted_count_ts, adjusted_count_ts2)
```

I think this PR is ready to be merged, but I am of course open to any suggestions or criticism. :wink: 
For instance, I am not sure if the naming of `adjust_timestamp` is correct. An alternative could be `base_timestamp` or `ref_timestamp` :thinking:?

Cheers,

----

- [X] closes #25226
closes #28302
closes #28675
closes #4197
closes #8521
- [X] Add 'origin' and 'offset' arguments to 'resample' and 'pd.Grouper'
- [X] tests added / passed
- [X] Add deprecation warning for `loffset` and `base` in the code
- [X] Add deprecation warning for `loffset` and `base` in the doc
- [x] Add examples in the doc for `origin` and `offset`
- [x] whatsnew entry (add deprecation notice with `offset` example)
- [X] passes `black pandas`
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

hasB4K committed:
ENH: add 'origin and 'offset' arguments to 'resample' and 'pd.Grouper'

pep8speaks commented:
Hello @hasB4K! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`pandas/core/resample.py`](https://github.com/pandas-dev/pandas/blob/fd3cc9d50526b1acd3fea0b2ceb46f2adf6b49a1/pandas/core/resample.py):

> [Line 1659:89](https://github.com/pandas-dev/pandas/blob/fd3cc9d50526b1acd3fea0b2ceb46f2adf6b49a1/pandas/core/resample.py#L1659): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (97 > 88 characters)

mroeschke commented:
Instead of adding a new keyword, might be nice if `base` could take a `Timestamp` instead since they are both relevant when a frequency is passed.

hasB4K commented:
I would suggest rather the following:

I always thought that the `base` argument was kind of ambiguous as a naming. And the current behavior is quite confusing. It needs to be an integer (or a floating point) that with the unit of the frequency.

- So `base=1` with a frequency of `5D` is equal to `1D` that we add as an offset.
- So `base=2` with a frequency of `5min` is equal to `2min` that we add as an offset.

This behavior is very confusing for the users (I include myself), but it also creates bugs: see #25226, #25161

Instead of relying on `base` I would rather deprecate this argument. The argument `loffset` (currently broken for pd.Grouper as shown by #28302, but fixable in the current PR) is kind of equivalent to what `base` is doing (especially since it is a Timedelta).

**Example of the current use of loffset with resample:**
```python
>>> start, end = "1/1/2000 00:00:00", "1/31/2000 00:00"
>>> rng = pd.date_range(start, end, freq="1231min")
>>> ts = pd.Series(np.arange(len(rng)), index=rng)
>>> ts.resample("1min", loffset=-pd.Timedelta("1min")).count()
```
```
1999-12-31 23:59:00    1
2000-01-01 00:00:00    0
2000-01-01 00:01:00    0
2000-01-01 00:02:00    0
2000-01-01 00:03:00    0
                      ..
2000-01-30 22:00:00    0
2000-01-30 22:01:00    0
2000-01-30 22:02:00    0
2000-01-30 22:03:00    0
2000-01-30 22:04:00    1
Freq: T, Length: 43086, dtype: int64
```

**Example of the current broken loffset argument:**
```python
>>> ts.groupby(pd.Grouper(freq="1min", loffset=-pd.Timedelta("1min"))).count()
```
```
2000-01-01 00:00:00    1
2000-01-01 00:01:00    0
2000-01-01 00:02:00    0
2000-01-01 00:03:00    0
2000-01-01 00:04:00    0
                      ..
2000-01-30 22:01:00    0
2000-01-30 22:02:00    0
2000-01-30 22:03:00    0
2000-01-30 22:04:00    0
2000-01-30 22:05:00    1
Freq: T, Length: 43086, dtype: int64
```

That being said, I agree that the naming of adjust_timestamp is not ideal. I would rename it into: `origin` or `base_timestamp`.

The line https://github.com/pandas-dev/pandas/blob/master/pandas/core/resample.py#L1728 would be replaced by something roughly equivalent to:
```
base_timestamp = start_of_day if base_timestamp is None else start_of_day
origin = base_timestamp.value + loffset.value
```

TL;DR:

1. I would fix in this PR `loffset` for pd.Grouper and deprecate the confusing `base` argument
2. I would rename the added argument `adjust_timestamp` into `origin` or `base_timestamp`
3. We would have `origin = base_timestamp.value + loffset.value`
4. I would add more tests to check the behavior of `loffset` and `base_timestamp`

hasB4K commented:
I just realised that `loffset` and `base` are not equivalent at all since this works:

```python
>>> start, end = "1/1/2000 00:00:00", "1/31/2000 00:00"
>>> rng = pd.date_range(start, end, freq="1231min")
>>> ts = pd.Series(np.arange(len(rng)), index=rng)
>>> ts.resample("1min", loffset=-pd.Timedelta("365D")).count()
```
```
1999-01-01 00:00:00    1
1999-01-01 00:01:00    0
1999-01-01 00:02:00    0
1999-01-01 00:03:00    0
1999-01-01 00:04:00    0
                      ..
1999-01-30 22:01:00    0
1999-01-30 22:02:00    0
1999-01-30 22:03:00    0
1999-01-30 22:04:00    0
1999-01-30 22:05:00    1
Freq: T, Length: 43086, dtype: int64
```

So I would suggest the following instead:

1. Deprecate the confusing `base` argument
2. Add an `origin_offset` argument that would be a Timestamp
3. Rename the added argument `adjust_timestamp` into `origin_timestamp`
4. We would have `origin_nanos = origin_timestamp.value + origin_offset.value`

I will not fix `loffset` in this PR since I am not sure of the behavior should be for pd.Grouper or how to fix it.

What do you think ?

jreback reviewed (changes requested):
I think base and loffset actually are pretty useful. However for non-evenly divisible freq the issue is that you likely simply want to use the first (or maybe the last) timestamp as the base. So how about we just add that ability in base to accept the string `first` or `last` rather than adding *another* keyword?

hasB4K commented:
@jreback this won't fix the issue that I'm trying to tackle. The idea is to be able to have a **fixed timestamp as a "origin"** that does not depend of the time series. So neither the base argument with `first` (which is the current behavior) or `last` string will fix the issue.

I could use the `base` argument and use it as the "origin" argument that I want to add if `base`is not a number like suggested @mroeschke. But I think this could create some confusion in the API (I still believe that `base` is useful but can be quite confusing to use).

jreback commented:
> @jreback this won't fix the issue that I'm trying to tackle. The idea is to be able to have a **fixed timestamp as a "origin"** that does not depend of the time series. So neither the base argument with `first` (which is the current behavior) or `last` string will fix the issue.
> 
> I could use the `base` argument and use it as the "origin" argument that I want to add if `base`is not a number like suggested @mroeschke. But I think this could create some confusion in the API (I still believe that `base` is useful but can be quite confusing to use).

@hasB4K not averse with changing things. But we currently have base, loffset, so I don' really like the idea of another another pretty opaque options. 

I would be onboard with deprecating both of these and replacing with 2 options, e.g. origin and offset come to mind.

hasB4K commented:
> I would be onboard with deprecating both of these and replacing with 2 options, e.g. origin and offset come to mind.

So would this signature be ok with you @jreback?
```
pd.Grouper(
    # ... arguments untouched with this enhancement like `freq`, `sort`, ...

    # ADDED arguments
    origin: pd.Timestamp, default None
        Only when freq is passed.
        The timestamp used as reference for the grouping bins.
        If None is passed, the first day of the time series at midnight is used.
    offset: pd.Timedelta, default None
        Only when freq is passed.
        An offset timedelta added to the origin.

    # DEPRECATED arguments
    base: int, default 0 (DEPRECATED)
    loffset: str, DateOffset, timedelta object (DEPRECATED)
)
```

jreback commented:
@hasB4K 

sure that looks reasonable.

we would need to have a pretty nice deprecation message that shows one how to convert base and/or loffset to the new args (as well as a whatsnew and warning box in the docs); they can bascially be the same though. its how we want folks to migrate.

hasB4K commented:
> sure that looks reasonable.

Perfect, I will implement that in this PR then :slightly_smiling_face: 

> we would need to have a pretty nice deprecation message that shows one how to convert base and/or loffset to the new args (as well as a whatsnew and warning box in the docs); they can bascially be the same though. its how we want folks to migrate.

Yep, it seems quite necessary! Is there an example of a nice deprecation message in the current (or in the old) code that I could look into?
For now, I was thinking of adding to the documentation of `resample` and `pd.Grouper` examples of "how to migrate". And in the code something like `this argument is deprecated, please see: <url>`.

jreback commented:
> Yep, it seems quite necessary! Is there an example of a nice deprecation message in the current (or in the old) code that I could look into?
> For now, I was thinking of adding to the documentation of `resample` and `pd.Grouper` examples of "how to migrate". And in the code something like `this argument is deprecated, please see: <url>`.

there are some (recently removed in 1.0.0) deprecation messages in resample on how to handle the freq arg. myabe not *great* but ok :->

hasB4K commented:
@jreback I still need to add more examples for 'origin' and 'offset' and update the "what's new" part of the doc, but otherwise, it's ready for review :slightly_smiling_face:

jreback commented on doc/source/user_guide/timeseries.rst:
```diff
@@ -1563,19 +1563,16 @@ end of the interval is closed:
-labels. ``label`` specifies whether the result is labeled with the beginning or
-the end of the interval. ``loffset`` performs a time adjustment on the output
-labels.
+Parameters like ``label`` are used to manipulate the resulting labels.
```
can you add a new section showing how to use origin (pretty much copy your whatsnew example and put it in the appropriate location). the idea is to have more full examples in the main docs and have a really simple example in whatsnew and point to the main docs.

hasB4K replied:
Yep, this is done in the last commit. "What's New" and "Timeseries" is up to date. :wink:

jreback commented on doc/source/whatsnew/v1.1.0.rst:
```diff
@@ -36,6 +36,30 @@ For example:
+Grouper now supports the argument origin
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+:class:`Grouper` and :class:`DataFrame.resample` now supports the argument `origin`. A the timestamp on which to adjust the grouping. (:issue:`31809`)
```
The timestamp

jreback commented on doc/source/whatsnew/v1.1.0.rst:
```diff
@@ -36,6 +36,30 @@ For example:
+
+   start, end = "1/1/2000 00:00:00", "1/31/2000 00:00"
+   rng = pd.date_range(start, end, freq="1231min")
+   ts = pd.Series(np.arange(len(rng)), index=rng)
```
show ts, then start a new ipython block

hasB4K replied:
Fixed in the last commit.

jreback commented on doc/source/whatsnew/v1.1.0.rst:
```diff
@@ -36,6 +36,30 @@ For example:
+   rng = pd.date_range(start, end, freq="1231min")
+   ts = pd.Series(np.arange(len(rng)), index=rng)
+   ts.groupby(pd.Grouper(freq="1399min")).agg("count")
+   ts.groupby(pd.Grouper(
```
I would show this example with .resample() as its simpler for people to get

hasB4K replied:
Fixed in the last commit.

jreback commented on pandas/core/generic.py:
```diff
@@ -7687,17 +7689,35 @@ def resample(
+            .. deprecated:: 1.1.0
+                You should add the loffset to the `df.index` after the resample.
+                like this:
+                ``df.index = df.index.to_timestamp() + to_offset(loffset)``
```
you can just say see below, no need to add the example here

hasB4K replied:
Fixed in the last commit.

jreback commented on pandas/core/generic.py:
```diff
@@ -7687,17 +7689,35 @@ def resample(
+
+            .. deprecated:: 1.1.0
+                The new arguments that you should use are 'offset' or 'origin'.
+                ``df.resample(freq="3s", base=2)``
```
same

hasB4K replied:
Fixed in the last commit.

jreback commented on pandas/core/generic.py:
```diff
@@ -7687,17 +7689,35 @@ def resample(
             For a MultiIndex, level (name or number) to use for
             resampling. `level` must be datetime-like.
+        origin : pd.Timestamp, default None
+            The timestamp on which to adjust the grouping. If None is passed,
```
add versionadded 1.1

hasB4K replied:
done in the last commit on every added argument on the public API.

jreback commented on pandas/core/generic.py:
```diff
@@ -7916,6 +7936,13 @@ def resample(
         2000-01-03     32     150
         2000-01-04     36      90
+
+        To replace the use of the deprecated loffset argument:
```
add an example using origin and offset

hasB4K replied:
Fixed in the last commit.

jreback commented on pandas/core/groupby/grouper.py:
```diff
@@ -68,9 +68,32 @@ class Grouper:
+        aggregated intervals. For example, for '5min' frequency, base could
+        range from 0 through 4. Defaults to 0.
+
+        .. deprecated:: 1.1.0
```
same comments as above

hasB4K replied:
Fixed in the last commit.

jreback commented on pandas/core/groupby/grouper.py:
```diff
@@ -68,9 +68,32 @@ class Grouper:
         Only when `freq` parameter is passed.
 
+        .. deprecated:: 1.1.0
+            loffset is only working for ``.resample(...)`` and not for
```
same

hasB4K replied:
Fixed in the last commit.

jreback commented on pandas/core/resample.py:
```diff
@@ -1266,7 +1269,7 @@ def get_resampler(obj, kind=None, **kwds):
     Create a TimeGrouper and return our resampler.
     """
-    tg = TimeGrouper(**kwds)
+    tg = TimeGrouper(**kwds, _warning_stack_level=2)
```
this is very odd to do, don't

hasB4K replied:
Yep! I agree, I changed that in the last commit, see https://github.com/pandas-dev/pandas/pull/31809#discussion_r379860446 for more context.

jreback commented on pandas/core/resample.py:
```diff
@@ -1357,20 +1365,51 @@ def __init__(
         self.limit = limit
-        self.base = base
+
+        self.origin = Timestamp(origin) if origin is not None else None
```
this must be tz aware if the resample index is right? where is this checked (and tested)? mention this in the doc-string.

hasB4K replied:
Yeah you are right, otherwise it could have worked but with potential confusion for the user.
This is resolved on my last commit. I have added a test on this case and a check is perform in `_get_timestamp_range_edges`.

jreback commented on pandas/core/resample.py:
```diff
@@ -1357,20 +1365,51 @@ def __init__(
+
         super().__init__(freq=freq, axis=axis, **kwargs)
 
+    @staticmethod
```
don't use static methods, you can create a module level function if you really need to, but I am not sure why you don't just in-line the warning (is it emitted in more than 1 place?)

hasB4K replied:
This should be resolved in the last commit.
> I am not sure why you don't just in-line the warning (is it emitted in more than 1 place?)

Well, yes and no... technically the TimeGrouper can raise alone the warnings. But the stack_level will be wrong in some cases depending if TimeGrouper is called from `.resample` (in `groupby.py` and `generic.py`) or from `pd.Grouper`... Here why I have added the nasty `_warning_stack_level` (a solution that I did not really liked to be honest)

Now I'm calling the function `_validate_resample_deprecated_args` from `__new__` of pd.Grouper or from the two `.resample` functions in `generic.py` or in `groupby.py`.

hasB4K replied:
Actually I found a way (but I don't know if I should do it or not) to `in-line` the warning. It can only be done in the `__new__` method of `pd.Grouper`. :man_shrugging: 

Depending if the `cls` is `TimeGrouper` (internal call of pandas) or `Grouper` I can infer what to do. Here a snippet of code that I just tested. Let me know if you prefer the manual call of `_validate_resample_deprecated_args` from `resample` or from `pd.Grouper` or this solution.

```python
class Grouper:

    def __new__(cls, *args, **kwargs):
        if kwargs.get("freq") is not None:
            from pandas.core.resample import TimeGrouper

            # if cls is TimeGrouper that means that the call comes from the
            # pandas internal of resample, otherwise it comes from pd.Grouper,
            # hence this test.
            warn_level = 4 if cls is TimeGrouper else 2
            if kwargs.get("base", None) is not None:
                warnings.warn(
                    "'base' in .resample() and in Grouper() is deprecated.\n"
                    "The new arguments that you should use are 'offset' or 'origin'.\n\n"
                    '>>> df.resample(freq="3s", base=2)\n'
                    "\nbecomes:\n\n"
                    '>>> df.resample(freq="3s", offset="2s")\n',
                    FutureWarning,
                    stacklevel=warn_level,
                )
                if kwargs.get("offset", None) is not None:
                    raise ValueError("offset and base cannot be present at the same time")

            if kwargs.get("loffset", None) is not None:
                warnings.warn(
                    "'loffset' in .resample() and in Grouper() is deprecated.\n"
                    "Here an example to have the same behavior than loffset:\n\n"
                    '>>> df.resample(freq="3s", loffset="8H")\n'
                    "\nbecomes:\n\n"
                    ">>> from pandas.tseries.frequencies import to_offset\n"
                    '>>> df = df.resample(freq="3s").mean()\n'
                    '>>> df.index = df.index.to_timestamp() + to_offset("8H")\n',
                    FutureWarning,
                    stacklevel=warn_level,
                )

            cls = TimeGrouper
        return super().__new__(cls)
```

jreback commented on pandas/core/resample.py:
```diff
@@ -1357,20 +1365,51 @@ def __init__(
+                "'base' in .resample() and in Grouper() is deprecated.\n"
+                "The new arguments that you should use "
+                "are 'offset' or 'origin'.\n\n"
+                '>>> df.resample(freq="3s", base=2)\n'
```
is there a case someone using base would actually need origin?

hasB4K replied:
Well, I thought that some people may use the base argument when they really want to use 'origin' instead. I mean, I wanted to use base with some nasty hack, but I'm doing this PR instead :sweat_smile:. But maybe that should not be in this warning message :wink: 

Let me know if you want me to remove the mention of this argument here!

jreback commented on pandas/core/resample.py:
```diff
@@ -1561,11 +1609,17 @@ def _get_period_bins(self, ax):
-        # GH 23882
-        if self.base:
+        # GH 23882 & 31809
+        if self.origin is not None or self.offset is not None or self.base:
```
can you simply remove all need for base (IOW convert it in the constructor)?

hasB4K replied:
Yep, it's done in the last commit!

jreback commented on pandas/core/resample.py:
```diff
@@ -1617,7 +1671,9 @@ def _take_new_index(obj, indexer, new_index, axis=0):
 
-def _get_timestamp_range_edges(first, last, offset, closed="left", base=0):
+def _get_timestamp_range_edges(
+    first, last, freq, closed="left", base=0, origin=None, offset=None
```
same here and all in this PR. remove the deprecated args, *except* from the user facing constructor. you need to fix the internal code to not use them

hasB4K replied:
Should be done in the last commit!

jreback commented on pandas/tests/resample/test_datetime_index.py:
```diff
@@ -789,11 +796,35 @@ def test_resample_base():
     tm.assert_index_equal(resampled.index, exp_rng)
 
 
+def test_resample_offset():
```
add the issue number as a comment

hasB4K replied:
Done in the last commit for all tests added.

jreback commented on pandas/tests/resample/test_period_index.py:
```diff
@@ -811,39 +812,80 @@ def test_resample_with_only_nat(self):
     @pytest.mark.parametrize(
-        "start,end,start_freq,end_freq,base",
+        "start,end,start_freq,end_freq,base,offset",
         [
```
remove base from all tests (except for a small number which by-definition are showing FutureWarning)

hasB4K replied:
I understand why you said that, but those tests are quite useful IMO. It avoids to break the base argument even if it is now deprecated. For instance to handle the conversion from `base` to `offset` in my last commit, those tests helped me.

I can delete those tests if you want to, but I was thinking: maybe I should just move them into a specific file like `resample/test_deprecated.py`? This file could be easily removed afterwards. :man_shrugging:

hasB4K replied:
I just did that in the last commit, I moved all deprecated tests into `test_deprecated.py`

jreback reviewed (changes requested):
loooking pretty good. a number of documenation & test comments. haven't looked at code changes too much. the main point is to remove the deprecated arguments *internaly* entirely. they should *only* appear in the user facing signatures. this should simplify the implementation.

hasB4K committed:
ENH: change how the warning are handled for base and loffset

hasB4K committed:
TST: move deprecated tests of loffset and base into test_deprecated

hasB4K committed:
ENH: check if origin for resample is timezone aware when needed

hasB4K commented:
@jreback I have changed the main code issues, I haven't changed the documentation yet. I'll do that probably tomorrow if I have some time :slightly_smiling_face:

hasB4K committed:
DOC: add 'Use  or  to adjust the start of the bins' section into timeseries.rst

hasB4K committed:
DOC: simplify doc for What's new and add a comment on the deprecation of loffset

hasB4K committed:
DOC: Add example for origin and offset in resample and in pd.Grouper

hasB4K commented:
@jreback The tests are passing, the doc has been updated, and I have fixed the issues in the code that you pinpointed. I think this PR is ready :slightly_smiling_face:

hasB4K commented:
By the way, there is some flaky tests not related to my PR (see https://github.com/pandas-dev/pandas/pull/32031). It appears since I have rebased with master.

The second reason why the CI can fail (at least that's the case this afternoon for what I have seed) comes from `continuous-integration/travis-ci/pr`. It seems that if fails when trying to upload the coverage :thinking: 

As shown here:
https://travis-ci.org/pandas-dev/pandas/jobs/651140009#L2423
```
...

= 61718 passed, 749 skipped, 603 xfailed, 7 xpassed, 211 warnings in 1238.98s (0:20:38) =
uploading coverage
bash <(curl -s https://codecov.io/bash) -Z -c -f /tmp/test_coverage.xml
/dev/fd/63: line 1: html: No such file or directory
/dev/fd/63: line 2: syntax error near unexpected token `<'
'
```

jreback commented on pandas/tests/resample/test_deprecated.py:
```diff
@@ -0,0 +1,264 @@
+from datetime import datetime, timedelta
+
+import numpy as np
```
can you ensure that we have replicated the loffset / base tests using offset / orgin. (this can be a followup as well). The idea is that when we enforce the deprecation we can simply remove the depredated tests and don't need to add any.

pls lmk either way.

hasB4K replied:
I have replicated the offset/origin tests:

- https://github.com/pandas-dev/pandas/pull/31809/files/689180c70a7833042a771795d78f8c667f5d3e54#diff-dd427a6bbdfef7e333ac320adb15614dR738
- https://github.com/pandas-dev/pandas/pull/31809/files/689180c70a7833042a771795d78f8c667f5d3e54#diff-dd427a6bbdfef7e333ac320adb15614dR738
- https://github.com/pandas-dev/pandas/pull/31809/files/689180c70a7833042a771795d78f8c667f5d3e54#diff-dd427a6bbdfef7e333ac320adb15614dR738
- https://github.com/pandas-dev/pandas/pull/31809/files/689180c70a7833042a771795d78f8c667f5d3e54#diff-967cd812d0942138129aacf0f9163604R82

however, I did not replicated the `loffset` argument since now it is up to the user to add the `loffset` to a `dataframe.index`

jreback replied:
so all tests that were above are now here?

> however, I did not replicated the loffset argument since now it is up to the user to add the loffset to a dataframe.index

what does this mean?

hasB4K replied:
> so all tests that were above are now here?

Yes, all deprecated tests have been moved into `test_deprecated.py`.
When you told me to remove some deprecated tests I moved them instead into `test_deprecated.py`: https://github.com/pandas-dev/pandas/pull/31809#discussion_r379711472

> > however, I did not replicated the loffset argument since now it is up to the user to add the loffset to a dataframe.index
> 
> what does this mean?

I have just moved the `loffset` tests into `test_deprecated.py` since neither of the two new arguments `origin` or `offset` really replace the now deprecated `loffset` argument.

`loffset` being just a timedelta that we add at the end of the resample, this behavior is just not handled with the new `origin` or `offset`.

To explain further, here a small notebook snippet (pretty much copy/paste from the new doc):

```python
>>> import pandas as pd
>>> import numpy as np
>>> from pandas.tseries.frequencies import to_offset
```

Setting up the timeseries and define a `my_loffset` for convenience
```python
>>> start, end = "2000-10-01 23:30:00", "2000-10-02 00:30:00"
>>> my_loffset = "19min"
>>> rng = pd.date_range(start, end, freq="7min")
>>> ts = pd.Series(np.arange(len(rng)) * 3, index=rng)
>>> ts
```
```
2000-10-01 23:30:00     0
2000-10-01 23:37:00     3
2000-10-01 23:44:00     6
2000-10-01 23:51:00     9
2000-10-01 23:58:00    12
2000-10-02 00:05:00    15
2000-10-02 00:12:00    18
2000-10-02 00:19:00    21
2000-10-02 00:26:00    24
Freq: 7T, dtype: int64
```
The old way:
```python
>>> ts_out = ts.resample("17min", loffset=my_loffset).sum()
>>> ts_out
```
```
2000-10-01 23:33:00     0
2000-10-01 23:50:00     9
2000-10-02 00:07:00    21
2000-10-02 00:24:00    54
2000-10-02 00:41:00    24
Freq: 17T, dtype: int64
```

The way to replace `loffset` after v1.1.0:
```python
>>> ts_out = ts.resample("17min").sum()
>>> ts_out.index = ts_out.index + to_offset(my_loffset)
>>> ts_out
```
```
2000-10-01 23:33:00     0
2000-10-01 23:50:00     9
2000-10-02 00:07:00    21
2000-10-02 00:24:00    54
2000-10-02 00:41:00    24
Freq: 17T, dtype: int64
```

jreback reviewed (changes requested):
looks pretty good. mostly cosmetic comments. ping on green.

jreback commented:
needs a rebase as well.

can you look thru open resample tests (I may have asked this already), and see if any examples are out there that this PR would close, if so pls add them as examples.

jreback commented:
cc @mroeschke if you'd have a look.

mroeschke commented on doc/source/user_guide/timeseries.rst:
```diff
@@ -1780,6 +1777,48 @@ natural and functions similarly to :py:func:`itertools.groupby`:
+
+.. ipython:: python
+
+    start, end = "1/10/2000 02:00:00", "1/20/2000 02:00"
```
Nit: Could you format the dates `YYYY-MM-DD ...`?

mroeschke commented on doc/source/user_guide/timeseries.rst:
```diff
@@ -1780,6 +1777,48 @@ natural and functions similarly to :py:func:`itertools.groupby`:
+
+    start, end = "1/10/2000 02:00:00", "1/20/2000 02:00"
+    middle = "1/15/2000 02:00"
+    rng = pd.date_range(start, end, freq="1231min")
```
Could we use a more digestible frequency example? e.g. `7 min` or something (and with the resampling frequency as well)

hasB4K replied:
Yeah, you are right, I did the change using 7min and 17min everywhere :)

mroeschke commented:
Looks like there's a diff with `doc/source/whatsnew/v1.0.0.rst` although it probably shouldn't have been touched. Could you double check the diff?

mroeschke commented on pandas/core/groupby/grouper.py:
```diff
@@ -68,9 +68,34 @@ class Grouper:
+            See: :class:`DataFrame.resample`
+
+    origin : pd.Timestamp, default None
+        The timestamp on which to adjust the grouping. It must be timezone
```
`It must be timezone aware if the index of the resampled data is.` -> `The timezone of the timestamp must match the timezone of the index`

hasB4K replied:
I did the change. However, I am not sure this entirely true since the `origin` just needs to be timezone aware when the index is. But I agree with you since it is less confusing for the reader anyway.

mroeschke commented on pandas/core/resample.py:
```diff
@@ -1593,6 +1614,34 @@ def _get_period_bins(self, ax):
+    if loffset is not None:
+        warnings.warn(
+            "'loffset' in .resample() and in Grouper() is deprecated.\n"
+            "Here an example to have the same behavior than loffset:\n\n"
```
Don't think `"Here an example to have the same behavior than loffset:\n\n"` adds much to this warning

mroeschke commented on pandas/core/resample.py:
```diff
@@ -1622,19 +1673,30 @@ def _get_timestamp_range_edges(first, last, offset, closed="left", base=0):
+        is_idx_tz_aware = first.tzinfo is not None or last.tzinfo is not None
+        if origin is not None and origin.tzinfo is None and is_idx_tz_aware:
+            raise ValueError(
+                "The origin must be timezone aware when the index "
```
`"The origin must have the same timezone as the index."`

hasB4K replied:
I did the change. However, I am not sure this entirely true since the `origin` just needs to be timezone aware when the index is. But I agree with you since it is less confusing for the reader anyway.

mroeschke commented on pandas/core/resample.py:
```diff
@@ -1675,12 +1737,16 @@ def _get_period_range_edges(first, last, offset, closed="left", base=0):
-    base : int, default 0
-        The "origin" of the adjusted Periods.
+    origin : pd.Timestamp, default None
+        The timestamp on which to adjust the grouping. It must be timezone
```
`It must be timezone aware if the index of the resampled data is.` -> `The timezone of the timestamp must match `the timezone of the index`

hasB4K replied:
Same comment: I did the change. However, I am not sure this entirely true since the `origin` just needs to be timezone aware when the index is. But I agree with you since it is less confusing for the reader anyway.

hasB4K commented:
> Looks like there's a diff with `doc/source/whatsnew/v1.0.0.rst` although it probably shouldn't have been touched. Could you double check the diff?

I just removed the 'execute' permission on this file, for some reason, this file was set as "executable". I thought that was a mistake since all others rst files didn't have this permission flag.

hasB4K commented on doc/source/whatsnew/v1.1.0.rst:
```diff
@@ -36,6 +36,55 @@ For example:
+    #   becomes:
+    ts.resample("2711min", offset="2min").sum()
+
+The argument ``loffset`` is now deprecated. (:issue:`31809`)
```
If I add `The argument ``loffset`` is now deprecated. You should now add an offset to the index DataFrame after being resampled` would that be okay?

jreback replied:
just add what you think is appropriate then we will comment

hasB4K commented on doc/source/user_guide/timeseries.rst:
```diff
@@ -1780,6 +1777,48 @@ natural and functions similarly to :py:func:`itertools.groupby`:
+   ts[middle:end].resample("2711min", origin=origin).sum()
+
+
+If needed you can just adjust the bins with an offset that would be added to the default ``origin``.
```
:thinking: - not sure to understand you here. Do you mean double backticks around `offset` ? `origin` cannot be a Timedelta, so I'm confused :sweat_smile:

jreback replied:
> Do you mean double backticks around offset

yes, anytime you use an actual kwarg you need to highlite;

mention that origin can be a Timestamp and offset a Timedelta

hasB4K commented on pandas/core/resample.py:
```diff
@@ -1593,6 +1614,34 @@ def _get_period_bins(self, ax):
+            '>>> df.resample(freq="3s", offset="2s")\n',
+            FutureWarning,
+            stacklevel=3,
+        )
```
Yes, they are: https://github.com/pandas-dev/pandas/pull/31809/files/689180c70a7833042a771795d78f8c667f5d3e54#diff-86037b7110970e90b82ba60bf8a9f011R40

hasB4K commented on pandas/core/groupby/grouper.py:
```diff
@@ -89,13 +114,85 @@ class Grouper:
         if kwargs.get("freq") is not None:
             from pandas.core.resample import TimeGrouper
+            from pandas.core.resample import _validate_resample_deprecated_args
+
```
Well. I wanted to avoid a circular imports here. Do you have a better solution ?
The real issue is that the warning needs to appear properly between resample AND pd.Grouper...

I explained the full issue (and an alternative solution) in an earlier comment of mine here:
https://github.com/pandas-dev/pandas/pull/31809#discussion_r379860446

So. What

jreback replied:
umm, don't validate here at all. you can simply validate all fo the args IN TimeGrouper itself. doing it here and there is so confusing.

hasB4K commented on pandas/core/resample.py:
```diff
@@ -1348,18 +1353,24 @@ def __init__(
+        self.offset = Timedelta(offset) if offset is not None else None
+        if base and isinstance(freq, Tick):
+            # this conversion handle the default behavior of base
+            # and the special case of GH #10530
```
Yes, to handle `base` argument when resampling TimedeltaIndex, `base` is treated as a pure offset. So with a frequency of `5s`, a base of `7` will be treated as an offset of 2s in the normal case (since we are doing a modulo on the frequency of `5s`), but it will treated as a pure offset (so `7s`) for TimedeltaIndex as you can see here: https://github.com/pandas-dev/pandas/commit/433602022bfbb9ee2424b77694c60f276065dee8#diff-e21e44fc4ab14efc2f09e0dacdd79fc6R241

hasB4K replied:
I just added a comment in the code to explain that

hasB4K reviewed (commented):
> can you look thru open resample tests (I may have asked this already), and see if any examples are out there that this PR would close, if so pls add them as examples.

@jreback not sure to understand here. Do you want to see if there is any **issues** that this PR would close? if so, the only two that I have found are #25226 and #28302. But maybe I missed some?

jreback commented on doc/source/user_guide/timeseries.rst:
```diff
@@ -1780,6 +1777,48 @@ natural and functions similarly to :py:func:`itertools.groupby`:
 See :ref:`groupby.iterating-label` or :class:`Resampler.__iter__` for more.
 
+.. _timeseries.adjust-the-start-of-the-bins:
+
```
can you do this

jreback commented on doc/source/user_guide/timeseries.rst:
```diff
@@ -1780,6 +1777,48 @@ natural and functions similarly to :py:func:`itertools.groupby`:
+Use `origin` or `offset` to adjust the start of the bins
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The bins of the grouping are adjusted based on the beginning of the day of the time series starting point. This works well with frequencies that are multiples of a day (like `30D`) or that divides a day (like `90s` or `1min`). But it can create inconsistencies with some frequencies that do not meet this criteria. To change this behavior you can specify a fixed timestamp with the argument ``origin ...
```
divides a day (add evenly)

jreback commented on doc/source/whatsnew/v1.1.0.rst:
```diff
@@ -36,6 +36,55 @@ For example:
+    start, end = "1/10/2000 02:00:00", "1/20/2000 02:00"
+    middle = "1/15/2000 02:00"
+    rng = pd.date_range(start, end, freq="1231min")
+    ts = pd.Series(np.arange(len(rng)) * 3, index=rng)
```
make this several blocks its *very* hard to grok this

hasB4K replied:
You are right... I tried to make this clearer in the new changes. Tell me if that's okay for you.

jreback commented on doc/source/whatsnew/v1.1.0.rst:
```diff
@@ -36,6 +36,55 @@ For example:
+
+.. ipython:: python
+
+    # ts.resample("2711min", base=2).sum()
```
don't use comments

jreback reviewed (changes requested):
pls address all of the comments before we look again

hasB4K commented on pandas/tests/resample/test_deprecated.py:
```diff
@@ -0,0 +1,264 @@
+        df.resample("3T", loffset="0s").sum()
+
+
+# old tests from test_base.py:
```
They were just there to emphasize that they could be removed later on. Just removed the comment in the changes that I will push shortly.

hasB4K committed:
CLN: review clean part one

hasB4K commented on pandas/core/generic.py:
```diff
@@ -7896,8 +7917,93 @@ def resample(
         2000-01-03     32     150
         2000-01-04     36      90
+
+        If you want to adjust the start of the bins based on a fixed timestamp:
```
I checked again to be sure, but it seems that I have already done that in an earlier commit.

hasB4K committed:
CLN: review clean part two

hasB4K committed:
DOC: update documentation to be more clearer (review part 3)

hasB4K commented:
Sorry for the delay, I have now more time to work on that PR :)
I think I have addressed all the points that needed changes.

There still the weird code on how I handle the warning for both `df.resample` and `pd.Grouper`. I explain more of this in these comments:
https://github.com/pandas-dev/pandas/pull/31809#discussion_r379861081 (full explanation of the issue)
https://github.com/pandas-dev/pandas/pull/31809#discussion_r387602206 (the follow-up comment)
https://github.com/pandas-dev/pandas/pull/31809#discussion_r387600949 (link of the tests of the warnings)

mroeschke reviewed (commented):
Looks like `doc/source/whatsnew/v1.0.0.rst` changed in the diff. Could you revert it?

hasB4K commented:
> Looks like `doc/source/whatsnew/v1.0.0.rst` changed in the diff. Could you revert it?

I removed the change. The only change I made there was 'chmod -x v1.0.0.rst' because, for some reason, this is the only 'whatsnew' doc file with the 'execute flag'. But anyway, this is not important :wink:

mroeschke commented:
> I removed the change. The only change I made there was chmod -x v1.0.0.rst because, for some reason, this is the only 'whatsnew' doc file with the 'execute flag'. But anyway, this is not important 😉

Ah I see. Happy to have that change in another PR

jreback commented on pandas/core/generic.py:
```diff
@@ -7974,8 +7995,85 @@ def resample(
+        2000-10-02 00:22:00    24
+        Freq: 17T, dtype: int64
+
+        >>> ts.resample("17min", origin=pd.Timestamp("1970-01-01")).sum()
```
hmm, how does this origin translate here? I actually find this confusing.

hasB4K replied:
@jreback Well origin is just a fixed point in time on which you decide to start the resampling. Using the epoch date was kind of handy in my mind:
* with the freq that divide or are a multiple of days: it works exactly in the same way
* with weird frequencies it's a fixed point in time kind of standard

But origin could have been any other date, we let the user decide what referential he wants to use.

jreback commented on pandas/core/resample.py:
```diff
@@ -1689,55 +1755,56 @@ def _get_period_range_edges(first, last, offset, closed="left", base=0):
     # error cause by resampling across multiple days when a one day period is
-    # not a multiple of the frequency.
-    #
-    # See https://github.com/pandas-dev/pandas/issues/8683
```
do you have tests from this issue?

hasB4K replied:
Yes, I just changed the comment to remove the link of the issue into GH 8683.
The test is untouched on my branch:

https://github.com/pandas-dev/pandas/blob/60353d5652d89bff9f2a71d05452bafe986f4c4c/pandas/tests/resample/test_datetime_index.py#L938

hasB4K replied:
That the same one from https://github.com/pandas-dev/pandas/pull/8941/files, but it seems that it has been moved (not by me) since 2014.

jreback commented on pandas/tests/resample/test_base.py:
```diff
@@ -195,35 +195,6 @@ def test_resample_empty_dtypes(index, dtype, resample_method):
         pass
 
 
-@all_ts
```
can you leave all of these tests. I am ok with either:
- converting to use offset / origin as appopriate, or simply catching the FutureWarning

hasB4K replied:
They have been moved into `test_deprecated.py`.
When you told me to remove some deprecated tests I moved them instead into `test_deprecated.py`: https://github.com/pandas-dev/pandas/pull/31809#discussion_r379711472

jreback commented on pandas/tests/resample/test_datetime_index.py:
```diff
@@ -412,70 +412,6 @@ def test_resample_frame_basic():
 
-@pytest.mark.parametrize(
-    "loffset", [timedelta(minutes=1), "1min", Minute(1), np.timedelta64(1, "m")]
-)
```
same as above

hasB4K replied:
They have been moved into `test_deprecated.py`.
When you told me to remove some deprecated tests I moved them instead into `test_deprecated.py`: https://github.com/pandas-dev/pandas/pull/31809#discussion_r379711472

jreback reviewed (changes requested):
docs look good. some comments. I am puzzled why you have the epoch date as the `origin`, can you clarify?

hasB4K committed:
CLN: review fix - move warning of 'loffset' and 'base' into pd.Grouper

hasB4K committed:
CLN: fix lint issue with isort

hasB4K commented:
I have fixed the main issue with the warning, even though I think this is clearly not a perfect solution.

Here a list of the last 3 ongoing discussion:
*  https://github.com/pandas-dev/pandas/pull/31809#discussion_r403509397 The solution to warn properly with the `stacklevel`
* https://github.com/pandas-dev/pandas/pull/31809#discussion_r403510811 The confusion of what happens with the deprecated `loffset` argument 
* https://github.com/pandas-dev/pandas/pull/31809#discussion_r403393179 The confusing example with `origin` and epoch

WillAyd commented on doc/source/user_guide/timeseries.rst:
```diff
@@ -1780,6 +1777,50 @@ natural and functions similarly to :py:func:`itertools.groupby`:
+
+.. ipython:: python
+
+    ts.resample("17min").sum()
```
Can you build the docs to check this if you haven't already? I *think* only the second expression gets printed here, though we probably want both to highlight the difference

hasB4K replied:
I have build the doc before committing this, here is what I see:

![timeseries_doc_1](https://user-images.githubusercontent.com/3902859/78950658-bd3e0a00-7acf-11ea-91bb-a9d7118697bd.png)

Since we see the output of this line, I think I can set this conversation as Resolve safely :slightly_smiling_face:

WillAyd commented on doc/source/user_guide/timeseries.rst:
```diff
@@ -1780,6 +1777,50 @@ natural and functions similarly to :py:func:`itertools.groupby`:
+.. ipython:: python
+
+   origin = pd.Timestamp("1970-01-01")
+   ts.resample("17min", origin=origin).sum()
```
Same comment here

hasB4K replied:
I have build the doc before committing this, here is what I see:

![timeseries_doc_2](https://user-images.githubusercontent.com/3902859/78950849-59681100-7ad0-11ea-8c92-516837fe16d8.png)

Since we see the output of this line, I think I can set this conversation as Resolve safely :slightly_smiling_face:

WillAyd commented on doc/source/user_guide/timeseries.rst:
```diff
@@ -1780,6 +1777,50 @@ natural and functions similarly to :py:func:`itertools.groupby`:
+
+.. ipython:: python
+
+    ts.resample("17min", origin=start).sum()
```
same comment

hasB4K replied:
I have build the doc before committing this, here is what I see:

![timeseries_doc_3](https://user-images.githubusercontent.com/3902859/78950905-861c2880-7ad0-11ea-8f50-cb23e1c84da4.png)

Since we see the output of this line, I think I can set this conversation as Resolve safely :slightly_smiling_face:

WillAyd replied:
Thanks!

WillAyd commented on pandas/core/generic.py:
```diff
@@ -7709,9 +7709,11 @@ def resample(
+        base: Optional[int] = None,
         on=None,
         level=None,
+        origin=None,
```
```suggestion
        origin: pd.Timestamp = None,
```

Ideally should annotate all new arguments

hasB4K replied:
Yeah this is a *very* good idea ! however I think we need to decide the proper type before doing that. More on this here: https://github.com/pandas-dev/pandas/pull/31809#discussion_r406554007

WillAyd commented on pandas/core/generic.py:
```diff
@@ -7709,9 +7709,11 @@ def resample(
         on=None,
         level=None,
+        origin=None,
+        offset=None,
```
```suggestion
        offset: Union[str, pd.Timedelta] = None,
```

I *think* this is what we need here; since you are wrapping this call in Timedelta might need to match whatever that constructor accepts, or explicitly require a Timedelta instance to be passed

hasB4K replied:
More on the same subject here: https://github.com/pandas-dev/pandas/pull/31809#discussion_r406554007

WillAyd commented on pandas/core/resample.py:
```diff
@@ -1314,7 +1316,9 @@ def __init__(
         kind=None,
         convention=None,
-        base=0,
+        base=None,
```
Can you annotate these?

WillAyd commented on pandas/core/resample.py:
```diff
@@ -1348,18 +1352,32 @@ def __init__(
         self.limit = limit
-        self.base = base
+
+        self.origin = Timestamp(origin) if origin is not None else None
```
Related to comment on annotations above, do we want to wrap things like this? Seems a little out of place to do here; I think should explicitly require Timestamp / Timedelta instances or pass through to other functions that will box accordingly

hasB4K replied:
I think what we need is to add something like this in `pandas/_typing.py`

(based on the documentation 
```python
# based on the doc and on pandas._libs.tslibs.timestamps.pyx::convert_to_tsobject
TimestampCompatibleType = Union[
    "Timestamp", datetime, np.datetime64, int, np.int64, float, str
]

# based on the doc and on pandas._libs.tslibs.timedeltas.pyx::convert_to_timedelta64
TimedeltaCompatibleType = Union[
    "Timedelta", timedelta, np.timedelta64, int, np.int64, float, str
]
```

And we will have in the constructor something like this
```python
origin: Optional[TimestampCompatibleType] = None,
offset: Optional[TimedeltaCompatibleType] = None,
```

The interest is really to handle origin/offset as string or any compatible type of Timestamp/Timedelta.

That kind of things would be supported out of the box (and I think this is what we want):
```python
ts.resample("17min", origin=dt.datetime(year=1993, month=1, day=5)).sum()
ts.resample("17min", origin="1993-01-05").sum()
ts.resample("17min", offset="16h30").sum()
```

I have started the implementation of this, I'll finish that soon. :wink:

hasB4K replied:
I have annotated the arguments `offset` and `origin` (in the last commit).

> Seems a little out of place to do here; I think should explicitly require Timestamp / Timedelta instances
To be honest I think doing the following is fine:

```python
self.origin = Timestamp(origin) if origin is not None else None
```
It is done elsewhere in the code. If I have to take only one example, here is a piece of code that handle the arguments of `pd.date_range`:

https://github.com/pandas-dev/pandas/blob/a6a1ab20e1fd3b9c54308f852a4427846b0695da/pandas/core/arrays/datetimes.py#L364

and the doc is quite similar:
```
start : str or datetime-like, optional
```

> or pass through to other functions that will box accordingly

I have checked the code of Timestamp, and is seems the constructor __new__  is acting like a "boxing" function since it does not return a copy of `ts_input` if `ts_input` is already a Timestamp:
https://github.com/pandas-dev/pandas/blob/a6a1ab20e1fd3b9c54308f852a4427846b0695da/pandas/_libs/tslibs/timestamps.pyx#L427

Same thing for Timedelta:
https://github.com/pandas-dev/pandas/blob/a6a1ab20e1fd3b9c54308f852a4427846b0695da/pandas/_libs/tslibs/timedeltas.pyx#L1232

It seems the related issues/PR are recent though :slightly_smiling_face: https://github.com/pandas-dev/pandas/issues/30543

WillAyd commented on pandas/tests/resample/test_deprecated.py:
```diff
@@ -0,0 +1,260 @@
+PERIOD_RANGE = (period_range, "pi", datetime(2005, 1, 1), datetime(2005, 1, 10))
+TIMEDELTA_RANGE = (timedelta_range, "tdi", "1 day", "10 day")
+
+all_ts = pytest.mark.parametrize(
```
This is neat; don't think we've done this elsewhere

hasB4K replied:
`all_ts` is a copy of `all_ts` from `pandas/tests/resample/test_base.py` :)
I just copied the test that will be deprecated in `test_deprecated.py` for an easier removal when we will reach pandas 2.0. :wink:

WillAyd commented on doc/source/user_guide/timeseries.rst:
```diff
@@ -1780,6 +1777,50 @@ natural and functions similarly to :py:func:`itertools.groupby`:
+.. ipython:: python
+
+    ts.resample("17min", origin=start).sum()
+    ts.resample("17min", offset=pd.Timedelta("23h30min")).sum()
```
Hmm any chance you gave thought to requiring a negative Timedelta here? I'm on the fence

hasB4K replied:
Yes, negative Timedelta works actually, but not as someone may think with a frequency of 17min (since its a prime number) on this example. Let me explain with some example:

```python
import pandas._testing as tm

start = pd.Timestamp("2000-10-01 23:30:00")
ts_out_1  = ts.resample("17min", origin=start).sum()
ts_out_1b = ts.resample("17min", offset="23h30m").sum()
tm.assert_series_equal(ts_out_1, ts_out_1b)
ts_out_1
```
```
2000-10-01 23:30:00     9
2000-10-01 23:47:00    21
2000-10-02 00:04:00    54
2000-10-02 00:21:00    24
Freq: 17T, dtype: int64
```

So, as expected, `ts_out_1` and `ts_out_1b` is identical. The cause is simple: the default origin is the start of the day at midnight of "2000-10-01" on this example (since it is the beginning of the timeseries)

Applying a negative delta here would only be identical if you do this:
```python
default_origin = start.normalize()  #  '2000-10-01 00:00:00'
ts_out_2  = ts.resample("17min", origin=default_origin - pd.Timedelta("30m")).sum()
ts_out_2b = ts.resample("17min", offset="-30m").sum()
tm.assert_series_equal(ts_out_2, ts_out_2b)
ts_out_2
```
```
2000-10-01 23:18:00     0
2000-10-01 23:35:00    18
2000-10-01 23:52:00    27
2000-10-02 00:09:00    39
2000-10-02 00:26:00    24
Freq: 17T, dtype: int64
```

But `ts_out_2` and `ts_out_1` are not equal:

```python
tm.assert_series_equal(ts_out_1, ts_out_2)
```
```
AssertionError: Series are different

Series length are different
[left]:  4, DatetimeIndex(['2000-10-01 23:30:00', '2000-10-01 23:47:00',
               '2000-10-02 00:04:00', '2000-10-02 00:21:00'],
              dtype='datetime64[ns]', freq='17T')
[right]: 5, DatetimeIndex(['2000-10-01 23:18:00', '2000-10-01 23:35:00',
               '2000-10-01 23:52:00', '2000-10-02 00:09:00',
               '2000-10-02 00:26:00'],
              dtype='datetime64[ns]', freq='17T')
```


BUT this is only the case when dealing with weird frequencies! If you take a more classical frequency (that divide or is a multiple of a day) this is true:

```python
ts_out_3 = ts.resample("15min", offset="23h30m").sum()
ts_out_3b = ts.resample("15min", offset="-30m").sum()
tm.assert_series_equal(ts_out_3, ts_out_3b)
ts_out_3
```
```
2000-10-01 23:30:00     9
2000-10-01 23:45:00    21
2000-10-02 00:00:00    33
2000-10-02 00:15:00    45
Freq: 15T, dtype: int64
```

WillAyd reviewed (changes requested):
Looks pretty good; I think still a few items around the API in terms of what it can accept to hash out, but very nice PR so far

hasB4K committed:
Update pandas/core/generic.py

Co-Authored-By: William Ayd <william.ayd@icloud.com>

jreback commented:
would it make sense for origin=‘epoch’ to possible? maybe even a future default?

hasB4K commented:
> would it make sense for origin=‘epoch’ to possible? maybe even a future default?

Yes it would, however setting that as a default would be a breaking change for frequencies that does not divide or is not a multiple of a day.

Currently `origin=0` works for setting epoch. Do you want me to add `'epoch'` as a possible value of `origin`?

hasB4K committed:
CLN: add TimestampCompatibleTypes and TimedeltaCompatibleTypes in pandas._typing

hasB4K committed:
CLN: fix lint issue with isort

hasB4K commented:
> would it make sense for origin=‘epoch’ to possible? maybe even a future default?

After investigating with #28675 and #8521, I think your comment raises a good remark. I think we should let `origin` be equal to:
* a timestamp (the current implementation)
* the string `'epoch'` which would be equal to the epoch (with a UTC tz)
* the string `'start'` which would set `origin` to the start of the timeseries (in `core.resample::_adjust_dates_anchored`, `origin` would be set to `first`) - it would solve issues like #28675 and #8521 more easily.
* the string `'start_day'` which would set the `origin` to the start of the day of the series. (the current default behavior)

I think we should let the default to `start_day` since not doing would be a breaking change.

@jreback what do you think?

jreback commented:
i think those are reasonable values

can you be sure the doc-string and examples show when to use

hasB4K committed:
ENH: support 'epoch', 'start_day' and 'start' for origin

hasB4K committed:
DOC: add doc for origin that uses 'epoch', 'start' or 'start_day'

hasB4K committed:
TST: add test for origin that uses 'epoch', 'start' or 'start_day'

hasB4K commented:
> i think those are reasonable values
> 
> can you be sure the doc-string and examples show when to use

@jreback I have added the support of `'epoch'`, `'start'` and `'start_day'` to the argument `origin` with its docs, examples and tests :slightly_smiling_face:. It should be ready for review.

hasB4K commented on pandas/core/resample.py:
```diff
@@ -1552,11 +1579,18 @@ def _get_period_bins(self, ax):
-        if self.base:
-            # get base adjusted bin edge labels
+        if isinstance(self.freq, Tick):
+            # GH 23882 & 31809: get adjusted bin edge labels with `origin`
+            # and `origin` support. This call only makes sense if the freq is a
+            # Tick since offset and origin are only used in those cases.
+            # Not doing this check could create an extra empty bin.
             p_start, end = _get_period_range_edges(
-                start, end, self.freq, closed=self.closed, base=self.base
+                start,
+                end,
+                self.freq,
+                closed=self.closed,
+                origin=self.origin,
+                offset=self.offset,
```
FWIW: I changed this:

```python
if self.origin is not None or self.offset is not None:
```

into

```python
if isinstance(self.freq, Tick):
```

When I realised that it makes only sense of doing this when freq was a Tick. However, I thought that this piece of code would have been quite similar, and that the `if` was done only for perfomance reason, but it seems that one `resample` test is failing if we always enter in the `if`: `pandas/tests/resample/test_period_index.py:210`.

```python
pandas/tests/resample/test_period_index.py:210 (TestPeriodIndex.test_resample_count[2M-expected_vals1])
self = <pandas.tests.resample.test_period_index.TestPeriodIndex object at 0x7f724082c520>
freq = '2M', expected_vals = [60, 40]

    @pytest.mark.parametrize(
        "freq,expected_vals", [("M", [31, 29, 31, 9]), ("2M", [31 + 29, 31 + 9])]
    )
    def test_resample_count(self, freq, expected_vals):
        # GH12774
        series = Series(1, index=pd.period_range(start="2000", periods=100))
        result = series.resample(freq).count()
        expected_index = pd.period_range(
            start="2000", freq=freq, periods=len(expected_vals)
        )
        expected = Series(expected_vals, index=expected_index)
>       tm.assert_series_equal(result, expected)
E       AssertionError: Series are different
E       
E       Series length are different
E       [left]:  3, PeriodIndex(['2000-01', '2000-03', '2000-05'], dtype='period[2M]', freq='2M')
E       [right]: 2, PeriodIndex(['2000-01', '2000-03'], dtype='period[2M]', freq='2M')
```

We have an empty bin that should not exists :man_shrugging:.

When I looked the code, it just seems that `_get_period_range_edges` did not handle well frequencies that are not Ticks.

hasB4K committed:
BUG: fix a timezone bug between origin and index on df.resample

hasB4K commented:
Just checking in, do I need to add or fix anything else @jreback? :slightly_smiling_face:

jreback commented on pandas/core/resample.py:
```diff
@@ -1346,18 +1351,35 @@ def __init__(
+        if origin in {"epoch", "start_day", "start"}:
+            self.origin = origin
+        else:
+            self.origin = Timestamp(origin)
```
you might want to change this sligthly
```
if origin in ....:
    self.origin = origin
else:
    try:
       self.origin = Timestamp(origin)
    except Exception as e:
        raise ValueError(....nice message here) from e
```

hasB4K replied:
I have added tests as long with a nice ValueError message for `origin` and `offset`

Here:
https://github.com/pandas-dev/pandas/blob/f2458dfaf5bdc22e22f58f7b720fa206b3cd88ef/pandas/tests/resample/test_datetime_index.py#L766

And here:
https://github.com/pandas-dev/pandas/blob/f2458dfaf5bdc22e22f58f7b720fa206b3cd88ef/pandas/tests/resample/test_datetime_index.py#L780

jreback commented on pandas/core/resample.py:
```diff
@@ -1622,49 +1658,60 @@ def _get_timestamp_range_edges(first, last, offset, closed="left", base=0):
-        if isinstance(offset, Day):
+    index_tz = first.tz
+    if isinstance(origin, Timestamp) and (origin.tz is None) != (index_tz is None):
+        raise ValueError("The origin must have the same timezone as the index.")
```
is there a test for this?

hasB4K replied:
Yes, here: https://github.com/pandas-dev/pandas/blob/f2458dfaf5bdc22e22f58f7b720fa206b3cd88ef/pandas/tests/resample/test_datetime_index.py#L817

jreback reviewed (changes requested):
looks pretty good. a few questions / suggestions. please rebase on master and we can likey merge this. ping when ready.

jreback commented:
@mroeschke if you can have a look.

jreback commented:
note that we are likey to merge: https://github.com/pandas-dev/pandas/pull/33498 first, so will need a rebase after.

mroeschke commented:
Could you add one test where where `resample("D")` is specified and we're resampling over a daylight savings time transition and an origin is specified? I know we have some specific code paths for this case and I'm curious about the behavior.

hasB4K committed:
DOC: change doc after review

hasB4K committed:
CLN: change typing for TimestampConvertibleTypes

hasB4K committed:
CLN: add nice message for ValueError of 'origin' and 'offset' in resample

hasB4K committed:
BUG: fix a bug when resampling in DST context

hasB4K commented on doc/source/whatsnew/v1.1.0.rst:
```diff
@@ -74,6 +74,72 @@ For example:
+
+.. ipython:: python
+
+    start, end = '2000-10-01 23:30:00', '2000-10-02 00:30:00'
```
It's done, let me know if the change is okay for you.

hasB4K commented on doc/source/user_guide/timeseries.rst:
```diff
@@ -1790,6 +1787,58 @@ natural and functions similarly to :py:func:`itertools.groupby`:
+.. ipython:: python
+
+    ts.resample('17min').sum()
+    ts[middle:end].resample('17min').sum()
```
What do you mean by "doesn't match"? :thinking:  The default behavior is 'start_day', so specifying it is not necessary. However, I have added `origin="start_day"` in one of my last commit to be more explicit about this example.

hasB4K commented on pandas/core/generic.py:
```diff
@@ -7762,17 +7766,40 @@ def resample(
+            must match the timezone of the index.
+            If a timestamp is not used, these values are also supported:
+
+            - 'epoch': `origin` is 1970-01-01
```
I have added tests as long with a nice message for `origin` and `offset`
Here:
https://github.com/pandas-dev/pandas/blob/f2458dfaf5bdc22e22f58f7b720fa206b3cd88ef/pandas/tests/resample/test_datetime_index.py#L766

And here:
https://github.com/pandas-dev/pandas/blob/f2458dfaf5bdc22e22f58f7b720fa206b3cd88ef/pandas/tests/resample/test_datetime_index.py#L780

hasB4K commented:
> ping when ready.

@jreback it should be ready now :wink:. I have answered to your earlier remarks/suggestions that you made, and did some fixes. Let me know if you need anything else.

> note that we are likey to merge: #33498 first, so will need a rebase after.

Okay :slightly_smiling_face: 

> Could you add one test where where `resample("D")` is specified and we're resampling over a daylight savings time transition and an origin is specified? I know we have some specific code paths for this case and I'm curious about the behavior.

@mroeschke You were right, I had a bug when dealing with a `resample("D")`. I fixed it and I added a bunch of tests here:
https://github.com/pandas-dev/pandas/blob/f2458dfaf5bdc22e22f58f7b720fa206b3cd88ef/pandas/tests/resample/test_datetime_index.py#L846

Let me know if you need anything else.

hasB4K committed:
TST: fix deprecation test

hasB4K committed:
TST: using pytz instead of datetutil in test of test_resample_origin_with_day_freq_on_dst

jreback commented:
ok #33498 was merged, pls merge master and ping on green.

hasB4K committed:
CLN: remove unused import

hasB4K commented:
@jreback Thank you for the merge of #33498! I rebased the current PR with master, let me know if you need anything else :slightly_smiling_face:

jreback commented:
very nice @hasB4K this was quite some PR! 

please have a read thru the built docs (https://dev.pandas.io/), will take a little bfeore they are there.

and if needed issue a followup to clarify.

and keep em coming!

hasB4K commented:
Thank you @jreback! :smiley:

The inputs and guidance from @mroeschke, @WillAyd and you was really interesting and challenging in the good way! I am really glad of the current state of this new functionality. Thank you all! :tada:
